private/bufpkg/bufcheck/bufcheckserver/internal/buflintvalidate/field.go

```diff
@@ -1037,6 +1014,104 @@ func checkExampleValues(
 	return nil
 }
 
+// createSyntheticFieldDescriptorForChecking creates a clone of a specific field
+// into its own message so that it can be checked independently. For map fields,
+// this includes creating a synthetic MapEntry as well.
+func createSyntheticFieldDescriptorForChecking(
+	// If specified, parentMapFieldDescriptor is the map field to be cloned,
+	// e.g. it is a field on a message that points to a map type.
+	parentMapFieldDescriptor protoreflect.FieldDescriptor,
+	// If parentMapFieldDescriptor is specified, this is a MapKey or MapValue.
+	// Otherwise, it is the non-map field to be cloned.
+	fieldDescriptor protoreflect.FieldDescriptor,
+	resolver protodesc.Resolver,
+) (
+	newParentMapFieldDescriptor protoreflect.FieldDescriptor,
+	newFieldDescriptor protoreflect.FieldDescriptor,
+	err error,
+) {
+	// We need to determine the "root" descriptor: if parentMapFieldDescriptor
+	// is set, then fieldDescriptor points to a MapKey or MapValue field, but we
+	// want to create a synthetic clone of the map field itself.
+	rootFieldDescriptor := fieldDescriptor
+	if parentMapFieldDescriptor != nil {
+		rootFieldDescriptor = parentMapFieldDescriptor
+	}
+	originalFile := rootFieldDescriptor.ContainingMessage().ParentFile()
+	newFileDescriptorProto := &descriptorpb.FileDescriptorProto{
+		Name:    proto.String(":synthetic:"),
+		Package: proto.String("__SYNTHETIC__"),
+		Syntax:  proto.String("proto3"),
+		Dependency: []string{
+			originalFile.Path(),
+		},
+	}
+
+	// Ensure that enum and message types are available to be resolved. It's a
+	// bit tricky to get this right, so we just import everything the original
+	// file containing the root message imports (and that file, too.)
+	for i := 0; i < originalFile.Imports().Len(); i++ {
+		newFileDescriptorProto.Dependency = append(
+			newFileDescriptorProto.Dependency,
+			originalFile.Imports().Get(i).Path(),
+		)
+	}
+
+	// Create a clone of the root field descriptor. Since our target message is
+	// going to be empty aside from this field, we also clear the OneofIndex.
+	newFieldDescriptorProto := protodesc.ToFieldDescriptorProto(rootFieldDescriptor)
+	newFieldDescriptorProto.OneofIndex = nil
+
+	newMessageDescriptorProto := &descriptorpb.DescriptorProto{
+		Name:  proto.String(string(rootFieldDescriptor.ContainingMessage().Name())),
+		Field: []*descriptorpb.FieldDescriptorProto{newFieldDescriptorProto},
+	}
+
+	// The map field and entry message have to be declared in the same scope, so
+	// we need to make a clone of the map entry type itself. This needs to be
+	// nested under the synthetic message.
+	if parentMapFieldDescriptor != nil {
+		// Shouldn't happen, but we don't want to panic if it does.
+		if !parentMapFieldDescriptor.IsMap() {
+			return nil, nil, errors.New("parentMapFieldDescriptor is not a map")
+		}
+		newMapEntryMessageDescriptor := &descriptorpb.DescriptorProto{
```

Comment:
I think this would be much safer (and less code) to instead use `protodesc` to mechanically translate the entire map entry message to a `*descriptorpb.DescriptorProto`, to make sure you don't miss anything here (such as features in an editions file).

Reply:
You'll also need to handle proto2 groups: in those cases, it should be fine to treat like any other message field, and let it refer to the group definition in the original file. But you'd need to change the field's type from `TYPE_GROUP` to `TYPE_MESSAGE`.